### PR TITLE
gas: fix the computation of `self.promises_gas`

### PR DESCRIPTION
**Pull request by nagisa**
With the wrapping computation of `new_used_gas` any contract is in full
control of this value, including those that are actually less than
either operand (due to overflows.)

This is not actually a big deal by itself, if not for the fact that it
gives an opportunity for the attacker to nuke the entire network out of
service.

The in-line comments largely explain the reasoning behind the fix and
why it preserves the protocol-facing behaviour, thus not necessitating a
protocol version change. To reiterate what the comments say, basically
the only case that we want to resolve is for when the attacker picks
a value of `new_used_gas` such that it is less than `burnt_gas`. Instead
of asserting and aborting we simply set `self.promises_gas = 0`.

---

The fix in this PR has been included in 1.29.3 (72d4a4d) and 1.30.0-rc.6 (4290758).

### CODE REVIEW

File: runtime/near-vm-logic/src/gas_counter.rs [hunk 1]
```diff
@@ -143,8 +145,21 @@ impl GasCounter {
...
+        // 1. `self.prepaid_gas < new_used_gas` This is perfectly fine and would be the happy path,
+        //    were the computations performed with arbitrary precision integers all the time.
+        // 2. `new_used_gas < new_burnt_gas` means that the `new_used_gas` computation wrapped
+        //    and `used_gas_limit` is now set to a lower value than it otherwise should be. In the
+        //    past this would have triggered an unconditional assert leading to nodes crashing and
+        //    network getting stuck/going down. We don’t actually need to assert, though. By
+        //    replacing the wrapping subtraction with a saturating one we make sure that the
+        //    resulting value of `self.promises_gas` is well behaved (becomes 0.) All a potential
+        //    attacker has achieved in this case is throwing some of their gas away.
+        self.promises_gas = used_gas_limit.saturating_sub(self.fast_counter.burnt_gas);
```
Comment (Ekleog-NEAR) [inline runtime/near-vm-logic/src/gas_counter.rs:162]: Given this is for the next release, do we want to take advantage of it to also fix the TODO just two lines above this diff, and just set self.promises_gas to 0 in the new protocol version? I’m thinking it’d allow simplifying quite a lot the comments, as the failure mode would become obvious.

File: runtime/near-vm-logic/src/gas_counter.rs [hunk 2]
```diff
@@ -127,7 +121,15 @@ impl GasCounter {
...
         } else {
-            Err(self.process_gas_limit(new_burnt_gas, new_burnt_gas + self.promises_gas).into())
+            // In the past `new_used_gas` would be computed using an implicit wrapping addition,
+            // which would then give an opportunity for the `assert` (now `debug_assert`) in the
+            // callee to fail, leading to a DoS of a node. A wrapping_add in this instance is
+            // actually fine, even if it gives the attacker full control of the value passed in
+            // here…
+            //
+            // [CONTINUATION IN THE NEXT COMMENT]
+            let new_used_gas = new_burnt_gas.wrapping_add(self.promises_gas);
```
Comment (Ekleog-NEAR) [inline runtime/near-vm-logic/src/gas_counter.rs:131]: For API clarity reasons, I’d argue for making this into a `checked_add` still if we go for the protocol change route: having a function for which one parameter is "literally whatever" sounds wrong to me. That said it definitely would require a protocol change, so… that’s only if we decide to do the todo listed below I guess!

**Review by jakmeier — APPROVED** [review]
I think we should merge this as-is to reach parity with the rc versions. Follow-up fixes can be discussed in separate issues.

**Comment by akhi3030** [comment]
> I think we should merge this as-is to reach parity with the rc versions. Follow-up fixes can be discussed in separate issues.

@jakmeier, @nagisa: what are the follow-up issues?  Did we create github issues for them?

**Comment by nagisa** [comment]
https://github.com/near/nearcore/issues/5148

**Pull request merged**